File: internal/domain/payment/payment.go
// Package payment contains domain entities for payment records.
// Payments map to tally.payment_head rows. Each payment belongs to one bill.
package payment

import (
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

// PayType represents the method of payment accepted by the system.
type PayType string

const (
	PayTypeCash     PayType = "cash"
	PayTypeWechat   PayType = "wechat"
	PayTypeAlipay   PayType = "alipay"
	PayTypeCard     PayType = "card"
	PayTypeCredit   PayType = "credit"
	PayTypeTransfer PayType = "transfer"
)

// validPayTypes contains all accepted PayType values.
var validPayTypes = map[PayType]struct{}{
	PayTypeCash:     {},
	PayTypeWechat:   {},
	PayTypeAlipay:   {},
	PayTypeCard:     {},
	PayTypeCredit:   {},
	PayTypeTransfer: {},
}

// Validate returns an error if p is not a recognised PayType.
func (p PayType) Validate() error {
	if _, ok := validPayTypes[p]; !ok {
		return fmt.Errorf("payment: invalid pay_type %q; must be one of cash/wechat/alipay/card/credit/transfer", p)
	}
	return nil
}

// Payment maps to one row in tally.payment_head.
// BillID maps to the related_bill_id FK column.
type Payment struct {
	ID          uuid.UUID       `json:"id"`
	TenantID    uuid.UUID       `json:"tenant_id"`
	BillID      uuid.UUID       `json:"bill_id"`
	PayType     PayType         `json:"pay_type"`
	Amount      decimal.Decimal `json:"amount"`
	TotalAmount decimal.Decimal `json:"total_amount"` // mirrors amount (no discount in MVP)
	PartnerID   *uuid.UUID      `json:"partner_id,omitempty"`
	OperatorID  *uuid.UUID      `json:"operator_id,omitempty"`
	CreatorID   uuid.UUID       `json:"creator_id"`
	BillNo      string          `json:"bill_no,omitempty"`
	PayDate     time.Time       `json:"pay_date"`
	Remark      string          `json:"remark,omitempty"`
	CreatedAt   time.Time       `json:"created_at"`
}

File: web/lib/api/payment.ts
/**
 * API wrapper for payment endpoints (Story 7.1).
 * Covers: record payment, list payments by bill.
 */
import { apiFetch } from "./client"
import type { Payment } from "./sale"

export type { Payment }

export const PAY_TYPE_LABEL: Record<string, string> = {
  cash: "现金",
  wechat: "微信",
  alipay: "支付宝",
  card: "银行卡",
  credit: "赊账",
  transfer: "转账",
}

export interface RecordPaymentRequest {
  bill_id: string
  amount: string
  payment_method: string
  remark?: string
}

export async function recordPayment(
  body: RecordPaymentRequest,
  tenantId?: string
): Promise<Payment> {
  return apiFetch<Payment>("/payments", { method: "POST", body: JSON.stringify(body), tenantId })
}

export async function listPayments(billId: string, tenantId?: string): Promise<Payment[]> {
  const result = await apiFetch<{ payments: Payment[] } | Payment[]>(
    `/payments?bill_id=${encodeURIComponent(billId)}`,
    { tenantId },
  )
  if (Array.isArray(result)) return result
  return (result as { payments: Payment[] }).payments ?? []
}
